examples/teams_flatten.cpp

```diff
@@ -0,0 +1,162 @@
+ *  simplifying indexing patterns
+ *
+ *  RAJA features shown:
+ *    - `launch' method to create kernel
+ *    -  flatten_block_threads
+ *
+ */
+
+/*
+ * Define device launch policies
+ */
+
+#if defined(RAJA_ENABLE_CUDA)
+using device_launch = RAJA::expt::LaunchPolicy<RAJA::expt::cuda_launch_t<false>>;
+using device_inner_loop0 =  RAJA::expt::LoopPolicy<RAJA::cuda_thread_x_direct>;
+using device_inner_loop1 =  RAJA::expt::LoopPolicy<RAJA::cuda_thread_y_direct>;
+using device_flatten_loop =  RAJA::expt::LoopPolicy<RAJA::expt::cuda_flatten_block_threads_xy_direct>;
+using reduce_policy = RAJA::cuda_reduce;
+#elif defined(RAJA_ENABLE_HIP)
+using device_launch = RAJA::expt::LaunchPolicy<RAJA::expt::hip_launch_t<false>>;
+using device_inner_loop0 =  RAJA::expt::LoopPolicy<RAJA::hip_thread_x_direct>;
+using device_inner_loop1 =  RAJA::expt::LoopPolicy<RAJA::hip_thread_y_direct>;
+using device_flatten_loop =  RAJA::expt::LoopPolicy<RAJA::expt::hip_flatten_block_threads_xy_direct>;
+using reduce_policy = RAJA::hip_reduce;
+#endif
+
+/*
+ * Define device launch policies
+ */
+
+using host_launch = RAJA::expt::LaunchPolicy<RAJA::expt::seq_launch_t>;
+using host_loop = RAJA::expt::LoopPolicy<RAJA::loop_exec>;
+
+int main(int argc, char *argv[])
+{
+
+#if defined(RAJA_ENABLE_CUDA) || defined(RAJA_ENABLE_HIP)
+
+  //
+  // Problem size dimensions
+  //
+  constexpr int N = 4;
+  constexpr int NN = N*N;
+
+  //
+  // Configure grid size
+  //
+  RAJA::expt::Grid grid(RAJA::expt::Teams(1),
+                        RAJA::expt::Threads(N, N),
+                        "Teams Flatten Kernel");
+
+  //
+  // Resource object for host, used to allocate memory
+  //
+  camp::resources::Host host_res;
+  int *h_A_ptr = host_res.allocate<int>(NN);
+
+  //
+  // Resource object for device, used to allocate memory
+  //
+#if defined(RAJA_ENABLE_CUDA)
+  camp::resources::Cuda device_res;
+#elif defined(RAJA_ENABLE_CUDA)
+  camp::resources::Hip device_res;
+#endif
+
+  int *d_A_ptr = device_res.allocate<int>(NN);
+
+//----------------------------------------------------------------------------//
+
+  std::cout << "\n Running device version of teams_flatten example ...\n";
+
+  RAJA::ReduceSum<reduce_policy, int> device_kernel_sum(0);
+  RAJA::View<int, RAJA::Layout<2>> d_A_2DView(d_A_ptr, N, N);
+  RAJA::View<int, RAJA::Layout<1>> d_A_1DView(d_A_ptr, NN);
+
+
+  RAJA::expt::launch<device_launch>
+    (grid, [=] RAJA_HOST_DEVICE (RAJA::expt::LaunchContext ctx)
+     {
+
+       RAJA::expt::loop<device_inner_loop1>(ctx, RAJA::RangeSegment(0, N), [&] (int j) {
+         RAJA::expt::loop<device_inner_loop0>(ctx, RAJA::RangeSegment(0, N), [&] (int i) {
+             d_A_2DView(j, i) = i + j;
+           });
+         });
+
+       ctx.teamSync();
+
+       // RAJA flatten policy will reshape a 2/3D thread team to 1D simplifying
+       // accumulating memory contents
+       RAJA::expt::loop<device_flatten_loop>(ctx, RAJA::RangeSegment(0, NN), [&] (int i) {
```

Comment:
Just trying to understand the usage of this flattening feature. If I wanted to use the `device_flatten_direct` policy, would I change this loop to look something like?

`RAJA::expt::loop<device_flatten_direct>(cxx, RAJA::RangeSegment(0,N), [&] (int j) {
     RAJA::expt::loop<device_flatten_direct>(cxx, RAJA::RangeSegment(0,N), [&] (int i) {`

Reply:
Nope, what this example shows how you can switch between using 2-level loops to 1-level loops on the GPU. The iteration space for the first two loops is [0, N] x [0, N] while the second is [0, NxN]. Since you have [NxN] threads you have to always having to specify a mapping for them, this just re-orders how things are mapped.